src/torchcodec/_core/BetaCudaDeviceInterface.cpp

```diff
@@ -0,0 +1,563 @@
+  // Callback setup, all are triggered by the parser within a call
+  // to cuvidParseVideoData
+  parserParams.pUserData = this;
+  parserParams.pfnSequenceCallback = pfnSequenceCallback;
+  parserParams.pfnDecodePicture = pfnDecodePictureCallback;
+  parserParams.pfnDisplayPicture = nullptr;
+
+  CUresult result = cuvidCreateVideoParser(&videoParser_, &parserParams);
+  TORCH_CHECK(
+      result == CUDA_SUCCESS, "Failed to create video parser: ", result);
+}
+
+// This callback is called by the parser within cuvidParseVideoData when there
+// is a change in the stream's properties (like resolution change), as specified
+// by CUVIDEOFORMAT. Particularly (but not just!), this is called at the very
+// start of the stream.
+// TODONVDEC P1: Code below mostly assume this is called only once at the start,
+// we should handle the case of multiple calls. Probably need to flush buffers,
+// etc.
+unsigned char BetaCudaDeviceInterface::streamPropertyChange(
+    CUVIDEOFORMAT* videoFormat) {
+  TORCH_CHECK(videoFormat != nullptr, "Invalid video format");
+
+  videoFormat_ = *videoFormat;
+
+  if (videoFormat_.min_num_decode_surfaces == 0) {
+    // Same as DALI's fallback
+    videoFormat_.min_num_decode_surfaces = 20;
+  }
+
+  if (!decoder_) {
+    decoder_ = NVDECCache::GetCache(device_.index()).getDecoder(videoFormat);
+
+    if (!decoder_) {
+      // TODONVDEC P0: consider re-configuring an existing decoder instead of
+      // re-creating one. See docs, see DALI.
+      decoder_ = createDecoder(videoFormat);
+    }
+
+    TORCH_CHECK(decoder_, "Failed to get or create decoder");
+  }
+
+  // DALI also returns min_num_decode_surfaces from this function. This
+  // instructs the parser to reset its ulMaxNumDecodeSurfaces field to this
+  // value.
+  return videoFormat_.min_num_decode_surfaces;
+}
+
+// Moral equivalent of avcodec_send_packet(). Here, we pass the AVPacket down to
+// the NVCUVID parser.
+int BetaCudaDeviceInterface::sendPacket(ReferenceAVPacket& packet) {
+  CUVIDSOURCEDATAPACKET cuvidPacket = {};
+
+  if (packet.get() && packet->data && packet->size > 0) {
+    // Regular packet with data
+    cuvidPacket.payload = packet->data;
+    cuvidPacket.payload_size = packet->size;
+    cuvidPacket.flags = CUVID_PKT_TIMESTAMP;
+    cuvidPacket.timestamp = packet->pts;
+
+    // Like DALI: store packet PTS in queue to later assign to frames as they
+    // come out
+    packetsPtsQueue.push(packet->pts);
+
+  } else {
+    // End of stream packet
+    cuvidPacket.flags = CUVID_PKT_ENDOFSTREAM;
+    eofSent_ = true;
+  }
+
+  CUresult result = cuvidParseVideoData(videoParser_, &cuvidPacket);
+  if (result != CUDA_SUCCESS) {
+    return AVERROR_EXTERNAL;
+  }
+  return AVSUCCESS;
+}
+
+// TODONVDEC P0: cleanup this raw pointer / reference monstruosity.
+ReferenceAVPacket* BetaCudaDeviceInterface::applyBSF(
+    ReferenceAVPacket& packet,
+    [[maybe_unused]] AutoAVPacket& filteredAutoPacket,
+    ReferenceAVPacket& filteredPacket) {
```

Comment:
Cleaning this up might be as simple as just returning a reference instead. And why does it accept `filteredAutoPacket`? I don't see any implementations that use it.